### PR TITLE
[SPARK-40813][CONNECT] Add limit and offset to Connect DSL

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

This PR decouples `LIMIT` and `OFFSET` to two plan which matches with DataFrame semantic where `OFFSET` is an independent operation and can happen before a LIMIT (or any other DataFrame).

### Why are the changes needed?

Improve the support for LIMIT and OFFSET.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT

### CODE REVIEW

**Comment by amaliujia** [comment]
R: @cloud-fan

**Comment by amaliujia** [comment]
cc @HyukjinKwon @zhengruifeng as I updated the python side accordingly.

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 1]
```diff
@@ -123,8 +123,16 @@ message Union {
 // Relation of type [[Fetch]] that is used to read `limit` / `offset` rows from the input relation.
 message Fetch {
```
Comment (HyukjinKwon) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:124]: qq why is this Fetch? There are two separate logical plans `Limit` and `Offset`.

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 2]
```diff
@@ -37,10 +37,11 @@ message Relation {
     Join join = 5;
     Union union = 6;
     Sort sort = 7;
-    Fetch fetch = 8;
+    Limit limit = 8;
     Aggregate aggregate = 9;
     SQL sql = 10;
     LocalRelation local_relation = 11;
+    Offset offset = 13;
```
Comment (grundprinzip) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:44]: Why two relations?

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 3]
```diff
@@ -173,10 +174,16 @@ class SparkConnectPlanner(plan: proto.Relation, session: SparkSession) {
...
+  private def transformLimit(limit: proto.Limit): LogicalPlan = {
     logical.Limit(
-      child = transformRelation(limit.getInput),
-      limitExpr = expressions.Literal(limit.getLimit, IntegerType))
+      limitExpr = expressions.Literal(limit.getLimit, IntegerType),
+      transformRelation(limit.getInput))
+  }
+
+  private def transformOffset(offset: proto.Offset): LogicalPlan = {
+    Offset(
```
Comment (grundprinzip) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:184]: please use `logical.Offset` to make sure where it comes from similar to the `proto.Offset`

**Review by grundprinzip — COMMENTED** [review]
Thanks for doing this. Maybe add some python tests as well? Or do you want to do this in a follow up?

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 2 above)
Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:44]: Please see the PR description. TLDR DataFrame does not follow SQL semantic.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 3 above)
Comment (cloud-fan) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:205]: +1

**Review by cloud-fan — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**